File: solutions/elixir/resistor-color-trio/1/lib/resistor_color_trio.ex
defmodule ResistorColorTrio do
  @doc """
  Calculate the resistance value in ohms from resistor colors
  """
  @spec label(colors :: [atom]) :: {number, :ohms | :kiloohms | :megaohms | :gigaohms}
  def label([first, second, magnitude | _]) do
    digits(first, second)
      |> apply_magnitude(magnitude)
      |> with_units()
  end

  defp digits(band_one, band_two), do: 10 * digit(band_one) + digit(band_two)

  defp apply_magnitude(ohms, magnitude), do: ohms * 10 ** digit(magnitude)

  defp digit(:black), do: 0
  defp digit(:brown), do: 1
  defp digit(:red), do: 2
  defp digit(:orange), do: 3
  defp digit(:yellow), do: 4
  defp digit(:green), do: 5
  defp digit(:blue), do: 6
  defp digit(:violet), do: 7
  defp digit(:grey), do: 8
  defp digit(:white), do: 9

  defp with_units(ohms) when ohms < 1_000, do: {ohms, :ohms}
  defp with_units(ohms) when ohms < 1_000_000,  do: {ohms / 1_000, :kiloohms}
  defp with_units(ohms) when ohms < 1_000_000_000, do: {ohms / 1_000_000, :megaohms}
  defp with_units(ohms), do: {ohms / 1_000_000_000, :gigaohms}
end

File: solutions/lua/resistor-color-trio/1/resistor-color-trio.lua
local convertOhms = function(ohms)
  local prefix = {'', 'kilo', 'mega', 'giga'}
  local index = 1
  while ohms > 0 and ohms % 1000 == 0 do
    ohms = ohms / 1000
    index = index + 1
  end
  return ohms, prefix[index] .. 'ohms'
end
local decode = function(c1, c2, c3)
    local resistances = {
      black = 0,
      brown = 1,
      red = 2,
      orange = 3,
      yellow = 4,
      green = 5,
      blue = 6,
      violet = 7,
      grey = 8,
      white = 9
    }
    
    local ohms = ((resistances[c1] * 10) + resistances[c2]) * 10^resistances[c3]
    return convertOhms(ohms)
end



return {
  decode = decode
}